neqo-http3/src/connection_client.rs

```diff
@@ -5930,4 +5938,113 @@ mod tests {
+        if conn.events().any(authentication_needed) {
+            conn.authenticated(AuthenticationStatus::Ok, now());
+        }
+    }
+
+    // Test that decoder stream type is always sent beofer any other instruction also
+    // in case when 0RTT is used.
+    #[test]
+    fn zerortt_request_use_dynamic_table() {
+        const MAX_TABLE_SIZE: u64 = 65536;
+        const MAX_BLOCKED_STREAMS: u16 = 5;
+        let mut client = default_http3_client_param(MAX_TABLE_SIZE);
+        let mut server = Http3Server::new(
+            now(),
+            DEFAULT_KEYS,
+            DEFAULT_ALPN_H3,
+            anti_replay(),
+            Rc::new(RefCell::new(FixedConnectionIdManager::new(5))),
+            QpackSettings {
+                max_table_size_encoder: MAX_TABLE_SIZE,
+                max_table_size_decoder: MAX_TABLE_SIZE,
+                max_blocked_streams: MAX_BLOCKED_STREAMS,
+            },
+        )
+        .unwrap();
+
+        let mut datagram = None;
+        let is_done = |c: &Http3Client| matches!(c.state(), Http3State::Connected);
+        while !is_done(&mut client) {
+            maybe_authenticate(&mut client);
+            datagram = client.process(datagram, now()).dgram();
+            datagram = server.process(datagram, now()).dgram();
+        }
+
+        // exchannge qpack settings, server will send a token as well.
+        datagram = client.process(datagram, now()).dgram();
+        datagram = server.process(datagram, now()).dgram();
+        let _ = client.process(datagram, now()).dgram();
+
+        let token = client
+            .events()
+            .find_map(|e| {
+                if let Http3ClientEvent::ResumptionToken(token) = e {
+                    Some(token)
+                } else {
+                    None
+                }
+            })
+            .unwrap();
+
+        // Make a new connection.
+        let mut client = default_http3_client_param(MAX_TABLE_SIZE);
+        assert_eq!(client.state(), Http3State::Initializing);
+        client
+            .enable_resumption(now(), &token)
+            .expect("Set resumption token.");
+
+        assert_eq!(client.state(), Http3State::ZeroRtt);
+        let zerortt_event = |e| matches!(e, Http3ClientEvent::StateChange(Http3State::ZeroRtt));
+        assert!(client.events().any(zerortt_event));
+
+        // Send a request using 0RTT data. This will trigger sending a HEADER_ACK on
+        // the decoder stream before stream id is written.
+        // The client will read HEADER_ACK (value: 0x80) as a stream type. To decode the stream
+        // type we will need 3 more byte (0x80 starts with 10 which is 4 byte encoded verint).
+        // 3 more bytes wil be the corect stream type (0x3) which will be added and to more
+        // requests that will trigger 2 HEADER_ACK. The client will cloed the stream and
+        // the server will trigger closing with error HttpClosedCriticalStream.
```

Comment:
Is this a comment about how this *used* to work?  Because what I see here is the request and response just working.

Reply:
Yes, I will update the comment